### PR TITLE
support temporary table preliminary (non-distributed, non-flush to s3)

### PR DESCRIPTION
**Pull request by jensenojs**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #6039

## What this PR does / why we need it:

### CODE REVIEW

**Review by reusee — APPROVED** [review]

**Comment by JackTan25** [comment]
@matrix-meow rerun all failed

File: pkg/frontend/mysql_cmd_executor.go [hunk 1]
```diff
@@ -2096,6 +2111,81 @@ func (cwft *TxnComputationWrapper) Compile(requestCtx context.Context, u interfa
 	if err != nil {
 		return nil, err
 	}
+	// check if it is necessary to initialize the temporary engine
+	if cwft.compile.NeedInitTempEngine(cwft.ses.InitTempEngine) {
```
Comment (daviszhen) [inline pkg/frontend/mysql_cmd_executor.go:2115]: add Get method for  `InitTempEngine` protected by mutex to remove the potential data race.

**Review by daviszhen — CHANGES_REQUESTED** [review]

File: pkg/frontend/mysql_cmd_executor.go (hunk 1 above)
Comment (jensenojs) [reply pkg/frontend/mysql_cmd_executor.go]: done

**Comment by jensenojs** [comment]
Need to double check for the visibility of temporary tables

**Review by daviszhen — APPROVED** [review]

**Review by yingfeng — APPROVED** [review]

File: pkg/frontend/mysql_cmd_executor.go [hunk 2]
```diff
@@ -2143,6 +2160,81 @@ func (cwft *TxnComputationWrapper) Compile(requestCtx context.Context, u interfa
...
+		if err != nil {
+			return nil, err
+		}
+
+		// 3. assign for compile, only for the first time needed
+		cwft.compile.SetTempEngine(e.TempEngine)
+		// 4. add auto_IncrementTable fortemp-db
+		colexec.CreateAutoIncrTable(e.TempEngine, requestCtx, cwft.proc, "temp-db")
+
+		txn := cwft.ses.txnCompileCtx.txnHandler
```
Comment (jensenojs) [inline pkg/frontend/mysql_cmd_executor.go:2234]: 1

File: pkg/frontend/mysql_cmd_executor.go [hunk 3]
```diff
@@ -2143,6 +2161,93 @@ func (cwft *TxnComputationWrapper) Compile(requestCtx context.Context, u interfa
...
+	if cwft.compile.NeedInitTempEngine(cwft.ses.IfNeedInitTempEngine()) {
+		// 0. init memory-non-dist engine
+		ck := clock.DefaultClock()
+		if ck == nil {
+			ck = clock.NewHLCClock(func() int64 {
+				return time.Now().Unix()
+			}, math.MaxInt)
+		}
+		shard := logservicepb.DNShardInfo{
+			ShardID:   2,
```
Comment (fengttt) [inline pkg/frontend/mysql_cmd_executor.go:2174]: What are these Shard/ReplicaId?   Why 2?

File: pkg/sql/compile/compile.go [hunk 4]
```diff
@@ -1112,16 +1140,33 @@ func (c *Compile) fillAnalyzeInfo() {
...
-	rel, err := db.Relation(c.ctx, n.TableDef.Name)
+	rel, err = db.Relation(c.ctx, n.TableDef.Name)
 	if err != nil {
-		return nil, err
+		var e error // avoid contamination of error messages
+		db, e = c.e.Database(c.ctx, "temp-db", c.proc.TxnOperator)
+		if e != nil {
+			return nil, e
+		}
+		rel, e = db.Relation(c.ctx, n.ObjRef.SchemaName+"-"+n.TableDef.Name)
```
Comment (fengttt) [inline pkg/sql/compile/compile.go:1159]: This seems to be a very bad idea.    How can you tell the different between schema   a, table b-c, and schema a-b, table c?

File: pkg/frontend/mysql_cmd_executor.go (hunk 3 above)
Comment (jensenojs) [reply pkg/frontend/mysql_cmd_executor.go:2174]: These parameters are fictitious and are not used for the memory citation engine.

File: pkg/sql/compile/compile.go (hunk 4 above)
Comment (jensenojs) [reply pkg/sql/compile/compile.go:1159]: will add test later

Comment (JackTan25) [reply pkg/sql/compile/compile.go:1159]: This is referenced to the initMemoryEngineNoDist function.

File: pkg/frontend/mysql_cmd_executor.go (hunk 3 above)
Comment (JackTan25) [reply pkg/frontend/mysql_cmd_executor.go:2174]: > What are these Shard/ReplicaId? Why 2?

This is referenced to the initMemoryEngineNoDist function.

**Review by aressu1985 — APPROVED** [review]

**Pull request closed without merging**